Extensions/MUC-Sub/XMPPMUCSub.m

```diff
@@ -0,0 +1,699 @@
+#endif
+
+#if DEBUG
+    static const int xmppLogLevel = XMPP_LOG_FLAG_TRACE; // | XMPP_LOG_FLAG_TRACE;
+#else
+    static const int xmppLogLevel = XMPP_LOG_LEVEL_WARN;
+#endif
+
+static NSString *const XMPPPubSubNamespace = @"http://jabber.org/protocol/pubsub#event";
+static NSString *const XMPPMUCSubNamespace = @"urn:xmpp:mucsub:0";
+static NSString *const XMPPMUCSubFeaturesPrefix = @"urn:xmpp:mucsub:nodes:";
+
+@interface XMPPMUCSub (PrivateAPI)
+
+@end
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+@implementation XMPPMUCSub
+
+- (id)init
+{
+    return [self initWithDispatchQueue:nil];
+}
+
+- (id)initWithDispatchQueue:(dispatch_queue_t)queue
+{
+    return [super initWithDispatchQueue:queue];
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+- (BOOL)activate:(XMPPStream *)aXmppStream
+{
+    XMPPLogTrace();
+    
+    if ([super activate:aXmppStream]) {
+        XMPPLogVerbose(@"%@: Activated", THIS_FILE);
+        
+        // xmppStream set by call to super.activate. moduleQueue set by super.init.
+        xmppIDTracker = [[XMPPIDTracker alloc] initWithStream:xmppStream dispatchQueue:moduleQueue];
+        return TRUE;
+    }
+    
+    return FALSE;
+}
+
+- (void)deactivate
+{
+    XMPPLogTrace();
+    
+    dispatch_block_t block = ^{ @autoreleasepool {
+        [xmppIDTracker removeAllIDs];
+        xmppIDTracker = nil;
+    }};
+    
+    if (dispatch_get_specific(moduleQueueTag)) {
+        block();
+    }
+    else {
+        dispatch_sync(moduleQueue, block);
+    }
+    
+    [super deactivate];
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+- (NSString *)supportedBy:(XMPPRoom *)room
+{
+    if (nil == room) {
+        return nil;
+    }
+    
+    // <iq from='hag66@shakespeare.example/pda'
+    //       to='coven@muc.shakespeare.example'
+    //     type='get'
+    //       id='ik3vs715'>
+    //   <query xmlns='http://jabber.org/protocol/disco#info'/>
+    // </iq>
+    
+    NSXMLElement *query = [NSXMLElement elementWithName:@"query" xmlns:XMPPDiscoInfoNamespace];
+    
+    NSString *iqId = [XMPPStream generateUUID];
+    
+    XMPPIQ *iq = [XMPPIQ iqWithType:@"get" elementID:iqId];
+    [iq addAttributeWithName:@"from" stringValue:xmppStream.myJID.bare];
+    [iq addAttributeWithName:@"to"   stringValue:room.roomJID.bare];
+    
+    [iq addChild:query];
+    
+    [xmppIDTracker addElement:iq target:self selector:@selector(handleSupportedByIQ:withInfo:) 
+                      timeout:60];
```

Comment:
What does XMPPIDTracker do with this? Were will the response go? (I'm sure it's down below somewhere, but I think a comment here would be helpful.)

Also, shouldn't the timeout be configurable, or use a timeout configured in XMPPFramework?

(Applies to several similar instances below.)

Reply:
Ohh, it calls the method given by `@selector`! Neat, never mind.

Reply:
The only defined timeout I could find was no timeout, waiting indefinitely. Other code I have seen used the hardcoded 60 seconds.

Reply:
Oh my ... it _is_ an ... interesting framework.